Windows Support and AppVeyor CI

ax3l opened the pull request:
Adds support for MSVC builds on Windows.

Due to various constrains in dependencies, this is
- Python 3 only
- HDF5 only
- Serial only (this is fine)

ADIOS1 on MSVC is very unlikely but ADIOS2 in the future could be a possibility, also for conda-forge.

Related to #63

## To Do

- [x] clean up verbose lines from experimenting on Windows
- [x] finalize PowerShell download script
- [x] README: Add Badge
- [x] tests: does not link with Catch2 enabled
- [x] tests: python bindings build, but do they run?
- [x] install? looks ok-ish, might need to add a link to the main `openPMD.dll` later on as well
- [x] enable AppVeyor in main repo as well

ax3l committed:
Download Samples: Powershell Script

Windows version of downloading all example files for tests.

ax3l committed:
AppVeyor: Windows CI

Add support for AppVeyor builds of serial openPMD-api with HDF5.

Dependencies are only available with C++11 support with VS2015
("[vc14]") toolchain. This is limited to Python 3 since Python 2.7
is only available with VS2008.

ax3l committed:
CMake: Windows Support

Paths, symbols and tests.

ax3l committed:
MSVC: long double == long

On MSVC, the range of `long double` is just the same as `long`.
This is fine, since it just needs to be "at least" double...

https://msdn.microsoft.com/en-us/library/s3f49ktz.aspx

https://en.wikipedia.org/wiki/Long_double

anokfireball commented on test/SerialIOTest.cpp:
```diff
@@ -815,7 +815,11 @@ TEST_CASE( "hdf5_dtype_test", "[serial][hdf5]" )
         s.setAttribute("vecFloat", std::vector< float >({0.f, 3.40282e+38f}));
         s.setAttribute("vecDouble", std::vector< double >({0., 1.79769e+308}));
+#ifdef _MSC_VER
+        s.setAttribute("vecLongdouble", std::vector< long double >({0.L, 1.49769e+308L}));
```
All of these calls could be replaced with [```std::numeric_limits::max```](http://en.cppreference.com/w/cpp/types/numeric_limits/max). This would make it compatible across all compilers (_that conform to the C++ standard_).

ax3l replied:
Yep, just a quick hack so far :)

ax3l commented:
@C0nsultant I think the tests are not compiling because of the redefinition of `private/protected` to `public` which causes binary incompatibilities in MSVC: https://github.com/catchorg/Catch2/issues/258#issuecomment-195479675

Can you maybe control these "very internal unit tests" via a define so I can skip them in MSVC?

anokfireball commented:
> Can you maybe control these "very internal unit tests" via a define so I can skip them in MSVC?

Just so I get this right - you have something like ```-DopenPMD_NONINVASIVE_TESTS=ON```  in mind, which then skip all tests that rely on ```#define public private```?
I'd rather have some modifier like [this](https://github.com/nlohmann/json/issues/913#issuecomment-357508906) that allows us to still test the data.

ax3l commented:
No, really the first one.

The second solution means we would ship different access modifiers on Windows which is a hack. Instead, let us just test *observable* behavior on Windows - that's ok.

anokfireball commented:
> let us just test observable behavior on Windows

Sighh...
The platform is not out main target audience anyway, so that should be fine. Core tests then can be designated for unix and assumed to represent all other platforms. Not elegant, but it does the job.

ax3l commented on CMakeLists.txt:
```diff
@@ -240,7 +249,23 @@ set_target_properties(openPMD PROPERTIES
+if(WIN32)
+    # generate a the openpmd_export.h header with export defines we can use for
+    # static members and other global data
+    include(GenerateExportHeader)
```
Comment out if we don't need this currently.

ax3l commented on CMakeLists.txt:
```diff
@@ -240,7 +249,23 @@ set_target_properties(openPMD PROPERTIES
+if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "MSVC")
+    target_compile_options(openPMD PUBLIC "/bigobj")
+    if(BUILD_SHARED_LIBS)
+        target_compile_options(openPMD PUBLIC "/MD")
```
check if we still need this block

ax3l commented on .appveyor.yml:
```diff
@@ -0,0 +1,140 @@
+  # Compiler & Generator Selection
+  - cmd: if "%APPVEYOR_BUILD_WORKER_IMAGE%"=="Visual Studio 2015" set OPENPMD_CMAKE_GENERATOR=Visual Studio 14 2015
+  - cmd: if "%TARGET_ARCH%"=="x64" set OPENPMD_CMAKE_GENERATOR=%OPENPMD_CMAKE_GENERATOR% Win64
+#  - cmd: if "%TARGET_ARCH%"=="x86" "C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\vcvarsall.bat" amd64_x86
```
note: I keep this for docs, this line is needed if we want to build everything from a `amd64` platform

ax3l committed:
README: Badge AppVeyor

anokfireball commented on test/SerialIOTest.cpp:
```diff
@@ -861,7 +864,9 @@ TEST_CASE( "hdf5_dtype_test", "[serial][hdf5]" )
     REQUIRE(s.getAttribute("vecFloat").get< std::vector< float > >() == std::vector< float >({0.f, 3.40282e+38f}));
     REQUIRE(s.getAttribute("vecDouble").get< std::vector< double > >() == std::vector< double >({0., 1.79769e+308}));
-    REQUIRE(s.getAttribute("vecLongdouble").get< std::vector< long double > >() == std::vector< long double >({0.L, 1.18973e+4932L}));
+#if !defined(_MSC_VER)
```
#179 is really this bad? Sad to see as ```long double``` is part of the C++ standard.

ax3l replied:
only during read :)
I think for some reason the variant takes the `double` alternative and then tries to read as `long double` or similar

anokfireball commented on test/SerialIOTest.cpp:
```diff
@@ -1171,7 +1176,9 @@ TEST_CASE( "adios1_dtype_test", "[serial][adios1]" )
     REQUIRE(s.getAttribute("float").get< float >() == 16.e10f);
     REQUIRE(s.getAttribute("double").get< double >() == 1.e64);
+#if !defined(_MSC_VER)
     REQUIRE(s.getAttribute("longdouble").get< long double >() == 1.e80L);
```
Why not use ```std::numeric_limits<long double>::max()``` here as well? Is it the initializer list causing the problems?

ax3l replied:
The problem I caused by the `long double` template argument during read #179 
I will exclude it for now and open a PR that removes the exclusion to demonstrate.

ax3l committed:
MSVC 2015: Disable Attribute Val

Do not print value of attribute that is read-only tried to
be set in exception. Work-around did only work on VS 2017 but
not on VS 2015.

Just throw a generic warning on MSVC instead.
